File: Python/ByRound/2064/2064_B_Variety_is_Discouraged.py
# Problem: CF 2064 B - Variety is Discouraged
# https://codeforces.com/contest/2064/problem/B

# ==============================================================================
# Problem: B. Variety is Discouraged
# Purpose: Maximize the score of an array after removing at most one contiguous subarray.
#          Score = length - number of distinct elements.
# Algorithms/Techniques:
#   - Two-pointer technique with frequency counting
#   - Sliding window to find optimal subarray to remove
# Time Complexity: O(n) per test case
# Space Complexity: O(n) for frequency array and input parsing
# ==============================================================================

for s in [*open(0)][2::2]:  # Read every second line starting from index 2 (skip t, n)
    n = len(a := [*map(int, s.split())])  # Convert current line to list of integers
    m = [0] * n  # Frequency array to track occurrences of each value (1 to n)
    for i in a:  # Count frequency of each element
        m[i - 1] += 1  # Adjust index to 0-based

    b = l = 0  # b = max segment length, l = start of current segment being considered
    r = [0, 0]  # r[0], r[1] = bounds of best subarray to delete
    
    for i in range(n):  # Iterate through all elements
        if m[a[i] - 1] != 1:  # If the element's count > 1 (i.e., it is repeated)
            l = i + 1  # Update start of potential segment with repeated elements
        elif b < i - l + 1:  # If current segment length is larger than best so far
            b = i - l + 1  # Update max length
            r = [l, i]  # Update bounds of segment to remove

    if b:  # If there was a valid segment found (non-zero length)
        print(r[0] + 1, r[1] + 1)  # Print in 1-based indexing
    else:
        print(0)  # No removal needed or impossible to improve score


# https://github.com/VaHiX/codeForces/

File: CPP/ByRound/2064/2064_B_Variety_is_Discouraged.cpp
// Problem: CF 2064 B - Variety is Discouraged
// https://codeforces.com/contest/2064/problem/B

/*
B. Variety is Discouraged

Purpose:
This program finds the optimal contiguous subarray to remove from an array such that the resulting array has maximum score,
where score = length - number of distinct elements.
The solution involves identifying a "single" element (appears only once in the current state), and extending
the longest contiguous sequence of such elements. Additionally, if we do not have any such sequence, we try removing
a subarray to maximize the score.

Algorithms:
- Use two sets: 'singles' for elements that appear exactly once, 'dupes' for elements that appear more than once.
- Scan through the array to categorize elements into singles and dupes, adjusting as we add new elements.
- Then scan again to find the longest contiguous subarray of elements from 'singles'.
- If there is no such subarray (or all are duplicates), no removal needed.

Time Complexity: O(n) per test case
Space Complexity: O(n) per test case

*/
#include <cstdio>
#include <set>
#include <vector>
int main() {
  long t;
  scanf("%ld", &t);
  while (t--) {
    long n;
    scanf("%ld", &n);
    std::vector<long> v(n);
    std::set<long> singles, dupes;
    for (long p = 0; p < n; p++) {
      scanf("%ld", &v[p]);
      if (dupes.count(v[p])) {
        continue; // Already marked as duplicate, skip
      } else if (singles.count(v[p])) {
        // First time seeing this element again -> now it's a duplicate
        dupes.insert(v[p]);
        singles.erase(v[p]);  // Remove from singles
      } else {
        singles.insert(v[p]); // New unique element
      }
    }
    long crl(-1), mxlen(-1);
    long left(-1), right(-1);
    for (long p = 0; p < n; p++) {
      if (singles.count(v[p])) {
        // Start or extend a valid segment
        if (crl < 0) {
          crl = p;
        }
        if (p - crl + 1 > mxlen) {
          mxlen = p - crl + 1;
          left = crl + 1;
          right = p + 1;
        }
      } else {
        // Reset segment
        crl = -1;
      }
    }
    if (left < 0) {
      puts("0");
    } else {
      printf("%ld %ld\n", left, right);
    }
  }
}


// https://github.com/VaHiX/codeForces/